tap_shopify/__init__.py

```diff
@@ -140,15 +141,32 @@ def sync():
         Context.state['bookmarks']['currently_sync_stream'] = stream_id
 
         with Transformer() as transformer:
-            for rec in stream.sync():
-                extraction_time = singer.utils.now()
-                record_schema = catalog_entry['schema']
-                record_metadata = metadata.to_map(catalog_entry['metadata'])
-                rec = transformer.transform(rec, record_schema, record_metadata)
-                singer.write_record(stream_id,
-                                    rec,
-                                    time_extracted=extraction_time)
-                Context.counts[stream_id] += 1
+            try:
+                for rec in stream.sync():
+                    extraction_time = singer.utils.now()
+                    record_schema = catalog_entry['schema']
+                    record_metadata = metadata.to_map(catalog_entry['metadata'])
+                    rec = transformer.transform(rec, record_schema, record_metadata)
+                    singer.write_record(stream_id,
+                                        rec,
+                                        time_extracted=extraction_time)
+                    Context.counts[stream_id] += 1
+            except pyactiveresource.connection.ResourceNotFound as exc:
+                raise ShopifyError(exc, 'Ensure shop is entered correctly') from exc
+            except pyactiveresource.connection.UnauthorizedAccess as exc:
+                raise ShopifyError(exc, 'Invalid access token - Re-authorize the connection') \
+                    from exc
+            except pyactiveresource.connection.ConnectionError as exc:
+                msg = ''
+                try:
+                    body_json = exc.response.body.decode()
+                    body = json.loads(body_json)
+                    msg = body.get('errors')
+                finally:
+                    raise ShopifyError(exc, msg) from exc
```

Comment:
Interesting, just to get this down, I wasn't sure this would work, but it seems like any exception thrown from JSON parsing will just get swallowed implicitly, and the true exception will bubble up. Nice.
```
>>> try:
...     raise Exception("1")
... except Exception as ex:
...     try:
...         raise Exception("2")
...     finally:
...         raise Exception("3") from ex
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
Exception: 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 7, in <module>
Exception: 3
```

Reply:
Yup, I was skeptical of it too but it works.